plugins/database/mongodb/connection_producer.go

```diff
@@ -117,21 +117,34 @@ func (c *mongoDBConnectionProducer) Connection(ctx context.Context) (interface{}
 		_ = c.client.Disconnect(ctx)
 	}
 
-	if c.clientOptions == nil {
-		c.clientOptions = options.Client()
+	vars := map[string]string{
+		"username": c.Username,
+		"password": c.Password,
 	}
-	c.clientOptions.SetSocketTimeout(1 * time.Minute)
-	c.clientOptions.SetConnectTimeout(1 * time.Minute)
-
-	var err error
-	opts := c.clientOptions.ApplyURI(c.ConnectionURL)
-	c.client, err = mongo.Connect(ctx, opts)
+	connURL := dbutil.QueryHelper(c.ConnectionURL, vars)
```

Comment:
Did you find that this actually changes something, e.g. `connURL` here is different than `c.ConnectionURL`? I _believe_ this is already done when the plugin is initialized with `Init()`:

- https://github.com/hashicorp/vault/blob/d1b6b354dcf92993fcc66e10f7f3ad0ad74bd594/plugins/database/mongodb/connection_producer.go#L83

Reply:
You're correct in that these two have a bit of a conflict, but actually it's the inverse of what you described. The `c.ConnectionURL = c.getConnectionURL()` call in `Init()` shouldn't be happening now. This call applies the username & password to the templated URL, meaning that if a user rotates the root password, the URL is using the old password because we haven't applied the new password to it because it's no longer templated.